buildSrc/src/main/java/org/elasticsearch/gradle/precommit/ValidateRestSpecTask.java

```diff
@@ -0,0 +1,160 @@
+ */
+
+package org.elasticsearch.gradle.precommit;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.networknt.schema.JsonSchema;
+import com.networknt.schema.JsonSchemaException;
+import com.networknt.schema.JsonSchemaFactory;
+import com.networknt.schema.SchemaValidatorsConfig;
+import com.networknt.schema.SpecVersion;
+import com.networknt.schema.ValidationMessage;
+import org.elasticsearch.gradle.util.Util;
+import org.gradle.api.file.FileTree;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.SkipWhenEmpty;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.util.PatternSet;
+import org.gradle.internal.Factory;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.StandardOpenOption;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Validates the REST specification against a schema. There should only be a single location for the schema, but any project can contribute
+ * to the REST specification.
+ */
+public class ValidateRestSpecTask extends PrecommitTask {
+
+    private static final String SCHEMA_PROJECT = ":rest-api-spec";
+    private static final String DOUBLE_STARS = "**"; // checkstyle thinks this is an empty javadoc statement, so string concat instead
+    private static final String JSON_SPEC_PATTERN_INCLUDE = DOUBLE_STARS + "/rest-api-spec/api/" + DOUBLE_STARS + "/*.json";
+    private static final String JSON_SPEC_PATTERN_EXCLUDE = DOUBLE_STARS + "/_common.json";
+    private static final String JSON_SCHEMA_PATTERN = DOUBLE_STARS + "/schema.json";
+    private static final ObjectMapper mapper = new ObjectMapper();
+    private Set<String> ignore = new HashSet<>();
+
+    @SkipWhenEmpty
+    @InputFiles
+    public FileTree getInputDir() {
+        // if null results in in NO-SOURCE
+        return Util.getJavaTestAndMainSourceResources(
+            getProject(),
+            Util.filterByPatterns(
+                getPatternSetFactory().create(),
+                Collections.singletonList(JSON_SPEC_PATTERN_INCLUDE),
+                Collections.singletonList(JSON_SPEC_PATTERN_EXCLUDE)
+            )
+        );
+    }
+
+    @Input
+    @Optional
+    public Set<String> getIgnore() {
+        return ignore;
+    }
+
+    public void ignore(String... ignore) {
+        this.ignore.addAll(Arrays.asList(ignore));
+    }
+
+    @Inject
+    protected Factory<PatternSet> getPatternSetFactory() {
+        throw new UnsupportedOperationException();
+    }
+
+    @OutputFile
+    public File getSuccessMarker() {
+        return new File(getProject().getBuildDir(), "markers/" + this.getName());
+    }
+
+    @TaskAction
+    public void validate() throws IOException {
```

Comment:
Bonus points if we make this task incremental so we only revalidate changed files 😉. This is a prime candidate for that kind of thing because each file can be independently validated.

https://docs.gradle.org/current/userguide/custom_tasks.html#incremental_tasks

Reply:
Done.  this is a neat feature I was not aware of.